### PR TITLE
ci: fix perms issue in label-pr workflow (round 2)

### PR DESCRIPTION
**Pull request by wexxlee**
#77 didn't quite do it, but this one should.  Tested against PRs from a forked repo, and it now works fine.

It's a little hacky in terms of a fix, but the reasons (and steps) are explained in a comment at the top of `label-pr-review.yml`.

### CODE REVIEW

File: .github/scripts/pr-review.cjs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    for (const pr of prs) {
+      const prNumber = pr.number;
+
+      // check if the PR has a `needs-review` label; if not, skip for efficiency
+      const prLabels = pr.labels;
+      const hasNeedsReviewLabel = prLabels
+        .map((label) => label.name)
+        .includes('needs-review');
+      if (!hasNeedsReviewLabel)
+        continue;
```
Comment (valarnin) [inline .github/scripts/pr-review.cjs:48]: Can we pre-filter for this label on the initial request instead? I'm not super familiar with the interfaces here, but a quick google suggests it might be possible?

File: .github/workflows/label-pr.yml [hunk 2]
```diff
@@ -64,21 +67,23 @@ jobs:
...
-          github.event.review.author_association == 'COLLABORATOR' ||
-          github.event.review.author_association == 'OWNER'
-        uses: buildsville/add-remove-label@v2.0.1
+      - uses: actions/checkout@v3
         with:
-          token: ${{secrets.GITHUB_TOKEN}}
-          labels: needs-review
-          type: remove
+          # force to ci checkout main repo to avoid
+          # unexpected PR code to be executed with out github token
```
Comment (valarnin) [inline .github/workflows/label-pr.yml:77]: The wording of this comment is a bit weird, maybe reword it to be more clear?

File: .github/workflows/label-pr.yml [hunk 3]
```diff
@@ -19,8 +19,11 @@ on:
       - auto_merge_enabled
       - auto_merge_disabled
       - ready_for_review
-  pull_request_review:
-    types: [submitted]
+  workflow_run:
+    workflows:
+      - PR Review
+    types:
+      - completed
```
Comment (valarnin) [inline .github/workflows/label-pr.yml:26]: It might be worth it to mention in a comment here that this is a workaround and to reference the `label-pr-review.yml` file for more details?

File: .github/workflows/label-pr.yml (hunk 2 above)
Comment (wexxlee) [reply .github/workflows/label-pr.yml:77]: It's actually an artifact from the old `labels.yaml` script that I didn't touch (and also appears in `lint-pr-title.yml`).  It is a little unclear, but I think the intent is to prevent the workflow from being executed against the head instead of base branch, which could allow someone to use the repo's secrets in newPR code, potentially exposing the token.  At least, I think that's the meaning?

File: .github/scripts/pr-review.cjs (hunk 1 above)
Comment (wexxlee) [reply .github/scripts/pr-review.cjs:48]: > Can we pre-filter for this label on the initial request instead? I'm not super familiar with the interfaces here, but a quick google suggests it might be possible?

I remember looking at this initially, and just did a little follow-up digging to confirm - but I don't believe the `Oktokit.js` interface to the GH REST API supports a `label` parameter, at least according to [the docs](https://octokit.github.io/rest.js/v20#pulls).  Do you happen to have a link for what you were looking at?

I'm reasonably sure you can pass a label value as a param if, e.g., you do a curl request and include it in the URL string, but, ehhh, I guess I'm not crazy about abandoning the Oktokit lib and figuring out how to handle the env vars, doing an http get, and then parsing the json?  Unless you had something different in mind?

Comment (valarnin) [reply .github/scripts/pr-review.cjs:48]: At the start of the docs:

> Custom request options can be passed as request.* options. See [@octokit/request](https://www.npmjs.com/package/@octokit/request#octokitrequest) options. The same options can be passed to each endpoint request method.

I believe that this implies that adding a `labels` parameter to the request should work (also based on other documentation for `@octokit/request` and the actual GH APIs)? e.g.:
```js
  const iterator = github.paginate.iterator(github.rest.pulls.list, {
    owner: owner,
    repo: repo,
    per_page: 100, // eslint-disable-line camelcase
    state: 'all',
    labels: 'needs-review',
  });
```

Comment (wexxlee) [reply .github/scripts/pr-review.cjs:48]: > At the start of the docs:
> 
> > Custom request options can be passed as request.* options. See [@octokit/request](https://www.npmjs.com/package/@octokit/request#octokitrequest) options. The same options can be passed to each endpoint request method.
> 
> I believe that this implies that adding a `labels` parameter to the request should work (also based on other documentation for `@octokit/request` and the actual GH APIs)? e.g.:
> 
> ```js
>   const iterator = github.paginate.iterator(github.rest.pulls.list, {
>     owner: owner,
>     repo: repo,
>     per_page: 100, // eslint-disable-line camelcase
>     state: 'all',
>     labels: 'needs-review',
>   });
> ```

Nice catch in the docs - didn't see that buried sentence.  But unfortunately, it doesn't seem to accept that param at least with pulls.list - tried with both `labels:` and `request: { labels: }` in my test script, and it's still yanking down all of the PRs.  But I'll keep poking at it.

Comment (wexxlee) [reply .github/scripts/pr-review.cjs:48]: > Nice catch in the docs - didn't see that buried sentence. But unfortunately, it doesn't seem to accept that param at least with pulls.list - tried with both `labels:` and `request: { labels: }` in my test script, and it's still yanking down all of the PRs. But I'll keep poking at it.

I've narrowed down at least part of the issue: the `pulls` endpoint doesn't support filtering by label (even though `issues` apparently does).  You can actually validate this right now at https://api.github.com/repos/OverlayPlugin/cactbot/pulls?labels=needs-review -- even with this param, it's still including PR 90, even though it doesn't have that label.

I've looked at a couple of the other methods, including `issuesAndPullRequests` and the general `search` method, but -- and I can't tell if this is just because of the way `actions/github` implemented the oktokit API or whether it's actually an oktokit limitation, they either don't play nicely with the pagination iterator or they don't return the full PR data (including reviews and commits) that the `pulls` endpoint does.  Open to figuring out a way to do this, but I'm running low on ideas...

Comment (wexxlee) [reply .github/scripts/pr-review.cjs]: > Can we pre-filter for this label on the initial request instead? I'm not super familiar with the interfaces here, but a quick google suggests it might be possible?

@valarnin Figured out a way to do this using a different method; tested and working now.

**Review by valarnin — APPROVED** [review]
